text/0011-utxo-model/0011-utxo-model.md

```diff
@@ -0,0 +1,194 @@
+
+``OutputID = Pair<Address, TransactionID>``
+
+All **colored balances** sent to the same address (by the same transaction) becomes part of the same output.
+
+### Colored Balances
+
+To be able to support tokenized assets on layer1 without the need for turing-complete smart contracts, we do not just
+use a numeric value to represent a balance, but instead allow balances to have an additional **color** which can be
+used to give coins a *meaning*.
+
+A **color** is simply a random sequence of bytes that can be set by the user and which is retained when being
+transferred:
+
+``Color = Array<byte>``
+
+Next to *user defined* colors, there are two *builtin color* values that carry a special meaning:
+
+- ``COLOR_IOTA = Array<byte>(0, 0, 0, ..., 0)`` represents the *base color* of uncolored coins.
+
+- ``COLOR_MINT = Array<byte>(255, 255, 255, ..., 255)`` represents a color which is being replaced by the
+  ``Transaction ID`` before being booked as an output. It is consequently used to "mint" new colored coins.
+
+A **colored balance** is accordingly a combination of a numeric balance with its corresponding color:
+
+``ColoredBalance = {Balance: uint64, Color: Color}``
+
+### Outputs
+
+An Output is a container for the colored balances that were spent by a particular transaction to a certain address. In
+addition to the list of colored balances and its ID, we store an ``OPCode`` and a corresponding ``OPCodeMetadata``.
+
+``Output = {ID: OutputID, Balances: Array<ColoredBalance>, OpCode: OPCode, OPCodeMetadata: OPCodeMetadata}``
+
+### OPCodes
+
+Currently, an output is *unlocked* if there is a valid signature for the address that holds the output, and it is not
+possible to define alternative conditions.
+
+There are however use cases where being able to define different conditions might be useful. A famous example are hash
+locks and time locks, that are used to create
+[Hashed Timelock Contracts](https://en.bitcoinwiki.org/wiki/Hashed_Timelock_Contracts) enabling things like atomic swaps
+and 2nd layer scaling solutions like the Lightning Network.
+
+The opcode is a single byte that is mapped to a specific unlock method. This will allow us to switch between different
+unlock conditions which introduces some very basic smart contract capabilities on the base layer.
+
+``OPCode = byte``
+
+Since some opcodes might require certain parameters (like a timestamp or a hash), we define a metadata field, that holds
+a marshaled version of these parameters.
+
+``OPCodeMetadata = Array<byte>`` 
+
+As long as these checks are not much more *expensive* than signature checks, it does not introduce problems but
+massively extends the functionality of the protocol. We have to however be very careful with adding new conditions, to
+not accidentally introduce potential bottlenecks.
+
+In contrast to bitcoin, we do not add a fully scriptable outputs (because that could result in long execution times),
+but we offer predefined methods for different use cases.
```

Comment:
What I understood is that by using `OP` + `OPCodeMetadata` instead of a script, you limit one op code use per output. We also don't have a stack.

Did I understand correctly?

About the long execution times, I guess that if we will use a script instead we can simply have a max length limit on it.
This would solve the problem no?

Even if it does, I don't know yet whether I am for scripts or not..
However, every time we would want a new functionality, people running old node version won't support it.
With scripting it will be more versatile, so we need to take this into account